git_theta/updates/base.py

```diff
@@ -9,40 +10,148 @@
     from importlib.metadata import entry_points
 
 import logging
-from typing import Optional
+from typing import Optional, Tuple
+
+import numpy as np
 
 from git_theta import git_utils, utils, params, metadata
 
 
-class Update:
-    """Base class for parameter update plugins."""
+Parameter = np.ndarray
 
-    def read(self, param_metadata):
-        lfs_pointer = param_metadata.lfs_metadata.lfs_pointer
-        serialized_param = git_utils.git_lfs_smudge(lfs_pointer)
-        param = params.get_update_serializer().deserialize(serialized_param)
-        return param
 
-    def get_last_version(self, repo, path, param_keys, param_metadata):
-        last_commit = param_metadata.theta_metadata.last_commit
-        logging.debug(f"Getting data from commit {last_commit}")
-        if last_commit:
-            last_metadata_obj = git_utils.get_file_version(repo, path, last_commit)
-            last_metadata = metadata.Metadata.from_file(last_metadata_obj.data_stream)
-            last_param_metadata = last_metadata.flatten()[param_keys]
-            return last_param_metadata
-        else:
-            raise ValueError("Cannot find previous version for parameters")
+class Update(metaclass=ABCMeta):
+    """Base class for parameter update plugins."""
+
+    def __init__(self, serializer: params.Serializer):
+        self.serializer = serializer
 
     @property
-    def name(self):
+    @abstractmethod
+    def name(self) -> str:
         raise NotImplementedError
 
-    def apply(self, repo, path, param_keys, param_metadata):
-        raise NotImplementedError
+    async def read(self, param_metadata: metadata.ParamMetadata) -> Parameter:
+        """Read in and deserialize a single parameter value based metadata."""
+        lfs_pointer = param_metadata.lfs_metadata.lfs_pointer
+        serialized_param = await git_utils.git_lfs_smudge(lfs_pointer)
+        param = (await self.serializer.deserialize(serialized_param))["parameter"]
+        return param
 
-    def calculate_update(self, repo, path, param_keys, param_metadata, param):
-        raise NotImplementedError
+    @abstractmethod
+    async def write(
+        self, param: Parameter, param_keys: Tuple[str], **kwargs
+    ) -> metadata.LfsMetadata:
+        """Serialize and save a parameter with git-lfs."""
+
+    @abstractmethod
+    async def apply(
+        self, param_metadata: metadata.ParamMetadata, param_keys: Tuple[str], **kwargs
+    ) -> Parameter:
+        """Get the final parameter value, including fetching previous values."""
+
+
+class TrueUpdate(Update):
```

Comment:
What is your reasoning for the name `TrueUpdate`? To me something like `RecursiveUpdate` or `IncrementalUpdate` that alludes to the idea that the update depends on previous values seems more descriptive.

Reply:
Updated to `IncrementalUpdate` good idea.